disambiguation: Author disambiguation part I

pazembrz opened the pull request:
*Added poetry to the project
*Removed invenio dependencies
*Upgraded to python 3
*Updated functions to use in memory dicts instead of files
*Added cli
*Added clustering per signature block

*INSPIRE-2601

michamos commented:
closing and reopening to trigger travis build (had to enable travis in the travis settings)

michamos commented on inspire_disambiguation/api.py:
```diff
@@ -21,127 +21,162 @@
+from inspire_disambiguation import conf
+from inspire_disambiguation.core.es.readers import get_signatures, get_input_clusters
+from inspire_disambiguation.core.ml.sampling import sample_signature_pairs
+from .core.ml.models import Clusterer, DistanceEstimator, EthnicityEstimator
```
Minor: why a relative import here but not for the others?

pazembrz replied:
We (me and Mona) are against relative import, We just missed this one.

drjova replied:
WDYM? Then we have a problem, we're using relative imports for modules in hep. We should discuss this IRL

pazembrz replied:
I'm not saying that it's wrong or something but we're creating something new. IMHO we should make it nice :)

michamos commented on inspire_disambiguation/core/es/readers.py:
```diff
@@ -0,0 +1,181 @@
+
+    Args:
+        signature_block (str): Signature block for which query should be
+            returned. If `empty` or `None` returns all of them.
```
```suggestion
            returned. If empty or `None` returns all of them.
```

michamos commented on inspire_disambiguation/core/es/readers.py:
```diff
@@ -0,0 +1,181 @@
+    query = (
+        LiteratureSearch()
+        .query(Q("bool", must=[literature_query, authors_query]))
+        .params(size=conf["ES_MAX_QUERY_SIZE"], _source=SIGNATURE_FIELDS)
```
do you need this `size` if you're doing a scan in any case?

pazembrz replied:
without `size` it will return 10 records per scan (If i remember correctly 10 is the default value). So with smaller number it will do more connections to ES.

michamos commented on inspire_disambiguation/app.py:
```diff
@@ -0,0 +1,143 @@
+        cls.config_data["LOG_LEVEL"] = "WARNING"
+        cls.config_data["REDIS_PHONETIC_BLOCK_KEY"] = "author_phonetic_blocks"
+        cls.config_data["REDIS_TIMEOUT"] = 60
+        cls.config_data["SAMPLED_PAIRS_SIZE"] = 12 * 50000
```
```suggestion
        cls.config_data["SAMPLED_PAIRS_SIZE"] = 12 * 100000
```
seems to be the value I was using previously

pazembrz replied:
:+1:

michamos commented on inspire_disambiguation/app.py:
```diff
@@ -0,0 +1,143 @@
+        cls.config_data["ES_TIMEOUT"] = 60
+        cls.config_data["ES_HOSTNAME"] = "localhost:9200"
+        cls.config_data["REDIS_URL"] = "redis://localhost:6379/0"
+        cls.config_data["ES_MAX_QUERY_SIZE"] = 9999
```
you're doing a scan, so is this the right value? is it needed at all?

pazembrz replied:
Discussed IRL

michamos commented on inspire_disambiguation/api.py:
```diff
@@ -21,127 +21,162 @@
+            clustered. If set to None, clustering will run on all blocks.
+    """
+    LOGGER.info(
+        "Pulling requested signatures and input_clusters ('%s') form ES",
```
```suggestion
        "Pulling signatures for block '%s' from ES",
```

drjova commented on inspire_disambiguation/app.py:
```diff
@@ -0,0 +1,143 @@
+
+    config_data = None
+    instance_path = os.path.dirname(os.path.abspath(__file__))
+    KEY_PREFIX = "DISAMBIGUATION_"
```
why we need key_prefix?

pazembrz replied:
key_prefix is only for values you read from env. And it's required so you won't load everything you exported in your shell

drjova commented on inspire_disambiguation/app.py:
```diff
@@ -0,0 +1,143 @@
+        cls.config_data["ES_HOSTNAME"] = "localhost:9200"
+        cls.config_data["REDIS_URL"] = "redis://localhost:6379/0"
+        cls.config_data["ES_MAX_QUERY_SIZE"] = 9999
+
```
why we don't use a normal config file for default values? or an ``.env`` file?

drjova replied:
for example https://github.com/theskumar/python-dotenv, there are so many standard ways to do configuration why we're choosing to implement ours?

pazembrz replied:
Normal config is still available in config.py. You can add there everything you need. This default values are here because they were here. and I didn't want to duplicate them there.

pazembrz replied:
This config was there, I only tweaked it a little. No one mentioned during refinement that it should be changed. No one requested it.

drjova replied:
man I'm not saying that wasn't here but we're creating something new. IMHO we should make it nice :) I'd move the default ones in ``config.py`` then, everything should be together

drjova commented on inspire_disambiguation/core/es/readers.py:
```diff
@@ -0,0 +1,181 @@
+    Example:
+        >>> get_signatures("ABDa", True)
+        [Signature(...), Signature(...), Signature(...)]
+
```
unecessary empty lines

drjova replied:
an in other files

drjova commented on inspire_disambiguation/core/helpers.py:
```diff
@@ -0,0 +1,47 @@
+    return get_recid_from_ref(author.get("record"))
+
+
+def get_recid_from_ref(ref_obj):
```
can we just reuse https://github.com/inspirehep/inspire-dojson/blob/master/inspire_dojson/utils/__init__.py#L87?

pazembrz replied:
No we cannot 
https://github.com/inspirehep/inspire-dojson/blob/bab0d6271f6262f362b4fb038ef2883bd9c50ec2/setup.py#L39

drjova commented on inspire_disambiguation/core/helpers.py:
```diff
@@ -0,0 +1,47 @@
+    return maybe_int(url.split("/")[-1])
+
+
+def _get_authors(record):
```
no need for ``_``

drjova replied:
```get_authors_full_names```?

drjova replied:
what kind of record?

drjova commented on inspire_disambiguation/core/helpers.py:
```diff
@@ -0,0 +1,47 @@
+    return get_value(author, "affiliations.value[0]", default="")
+
+
+def _get_author_id(author):
```
no nedd for ``_``

drjova commented on inspire_disambiguation/core/helpers.py:
```diff
@@ -0,0 +1,47 @@
+
+
+def _get_author_affiliation(author):
+    return get_value(author, "affiliations.value[0]", default="")
```
no need for ``_``

drjova commented on README.rst:
```diff
@@ -0,0 +1,26 @@
+    Copyright (C) 2014-2019 CERN.
+
+    INSPIRE is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
```
small detail I'd go for ``md`` since we're using it in other repos as well

drjova commented on inspire_disambiguation/api.py:
```diff
@@ -21,127 +21,162 @@
+                    ethnicity_model_path, distance_model_path, n_jobs, signature_block
+                ),
+                indent=4,
+            ),
```
multiline logs don't look good on ES at the moment, fyi

pazembrz replied:
This is only temporary until disambiguation part2 will be finished. I'm not sure, but I believe that until then all runs of this script will be done manually so readable output is quite important.

drjova commented on inspire_disambiguation/__init__.py:
```diff
@@ -21,11 +21,18 @@
+root.setLevel(getattr(logging, conf["LOG_LEVEL"]))
+
+handler = logging.StreamHandler(sys.stdout)
+handler.setLevel(getattr(logging, conf["LOG_LEVEL"]))
```
😿

drjova replied:
we don't need all this dance IMO:

```python
logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=conf["LOG_LEVEL"])  
```

michamos commented on inspire_disambiguation/api.py:
```diff
@@ -21,127 +21,162 @@
+        for pair in sample_signature_pairs(
+            curated_signatures, input_clusters, sampled_pairs_size
+        )
+    ]
```
maybe it's clearer with `list(sample_signature_pairs(curated_signatures, input_clusters, sampled_pairs_size))`

pazembrz replied:
IMO current version is more clear, but what @monaawi thinks?

drjova commented on inspire_disambiguation/core/es/readers.py:
```diff
@@ -0,0 +1,181 @@
+        )
+
+
+def get_lit_records_query(signature_block, only_curated):
```
```suggestion
def get_literature_records_query(signature_block, only_curated):
```

drjova commented on inspire_disambiguation/core/es/readers.py:
```diff
@@ -0,0 +1,181 @@
+            records (with `authors.curated_relation` set to `True`).
+
+    Returns:
+        Query: Query built for provided parameters
```
```suggestion
        Query: Query built for provided parameters.
```

drjova commented on inspire_disambiguation/core/es/readers.py:
```diff
@@ -0,0 +1,181 @@
+        else:
+            signatures_without_author.append(signature["signature_uuid"])
+    for cluster_id, (author_id, signature_uuids) in enumerate(
+        six.iteritems(signatures_with_author)
```
aren't we using python3 only? why we need ``six``?

pazembrz replied:
Yup, you are right. I forgot to update this old code.

michamos commented on inspire_disambiguation/core/ml/models.py:
```diff
@@ -319,68 +585,118 @@ def __init__(self, estimator):
-        signatures_by_uuid = load_signatures(signatures_path, publications_path)
+        Args:
+            signatures (iterable): Signatures which should be processed
+            input_clusters (iterable): TODO
```
wrong docstring

drjova commented on run-tests.sh:
```diff
@@ -20,6 +20,5 @@
-
-from __future__ import absolute_import, division, print_function
+poetry run flake8 &&
+poetry run py.test tests
```
end of the file

michamos commented on inspire_disambiguation/core/ml/models.py:
```diff
@@ -319,68 +585,118 @@ def __init__(self, estimator):
 
 @attr.s(slots=True)
 class Signature(object):
+    """Signature model for estimators"""
```
I wouldn't call it a model (at least it's not a ML model), it's a data structure, maybe even move it out to a separate file?

michamos commented on inspire_disambiguation/core/ml/models.py:
```diff
@@ -319,68 +585,118 @@ def __init__(self, estimator):
+                "author_affiliation": _get_author_affiliation(author),
+                "author_id": author_id,
+                "author_name": author["full_name"],
+                "publication": Publication.build(record),
```
If I remember correctly, previously the value was shared among all signatures in a given publication. Here you are creating a new publication for every signature, which might us significantly more memory.

michamos commented on inspire_disambiguation/core/ml/models.py:
```diff
@@ -319,68 +585,118 @@ def __init__(self, estimator):
 
 @attr.s(slots=True)
 class Publication(object):
+    """Publication model for estimators"""
```
same comment about `model`

drjova commented on tests/integration/test_models.py:
```diff
@@ -0,0 +1,357 @@
+    input_clusters = get_input_clusters(signatures)
+    clusterer = Clusterer(distance_estimator_mock)
+    clusterer.load_data(signatures, input_clusters)
+    expected_X = array(
```
why ``X`` uppercase and ``y`` lowercase?

pazembrz replied:
It's `sklearn` convention.

drjova commented on tests/unit/test_cli.py:
```diff
@@ -0,0 +1,102 @@
+def test_distance_model_cli_function_default_params(train_and_save_distance_model_mock):
+    runner = CliRunner()
+    runner.invoke(cli, ["train", "distance-model"])
+    train_and_save_distance_model_mock.assert_called_with(
```
what's the point of these tests? just to make sure it runs the proper functions?

pazembrz replied:
To be sure that proper function receives proper parameters.

michamos commented on pyproject.toml:
```diff
@@ -0,0 +1,46 @@
+
+
+[tool.poetry.dependencies]
+python = ">=3.6,<3.7"
```
why not python 3.7?

michamos commented on inspire_disambiguation/core/helpers.py:
```diff
@@ -0,0 +1,47 @@
+
+
+def _get_authors(record):
+    return get_value(record, "authors.full_name", default=[])
```
there's a whole list of similar functions in `models.py`. Maybe worth putting everything in the same place?

pazembrz replied:
Functions in models.py are used explicitly by `DistanceEstimator` when it's building this `FeatureUnion` Other helpers are used in different places. I don't want to mix things which are specifically for Estimators with other helpers.

michamos commented on inspire_disambiguation/cli.py:
```diff
@@ -0,0 +1,135 @@
+    click.secho(f"Done. Model saved in {save_model_path}", fg="green")
+
+
+# This needs to change to read from redis and cluster
```
leftover?

pazembrz replied:
:+1:

michamos commented on inspire_disambiguation/config.py:
```diff
@@ -21,20 +21,3 @@
 # or submit itself to any jurisdiction.
 
 """Disambiguation configuration."""
```
what is this file used for now?

pazembrz replied:
~You can put config variables there as things in app.py shouldn't be touched.~
It will contain all config variables.

michamos commented on inspire_disambiguation/api.py:
```diff
@@ -21,127 +21,162 @@
+        )
+        if not signature_block_data:
+            LOGGER.warning("No signature blocks in redis to process! STOP.")
+            break
```
This is not working like the continuous migration, as it will finish once the queue is empty. It's probably better to do it like there and run forever (maybe in a different PR).

michamos reviewed (commented):
it looks like you're avoiding writing to files, which will result in many more things kept in memory. You're also doing more copies than previously (`publications` inside signatures). Not sure if it won't blow up the RAM.

michamos commented on inspire_disambiguation/api.py:
```diff
@@ -21,127 +21,162 @@
+    return process_clustering_output(clusterer)
+
 
+def process_clustering_output(clusterer):
```
maybe move this somewhere else? it's not part of the API.

pazembrz replied:
I believe we can do this when output will be more stable.

michamos commented on inspire_disambiguation/core/helpers.py:
```diff
@@ -0,0 +1,350 @@
+        Object which will be cached must have some kind of factory.
+
+        Args:
+            identifier: Some uniquer keys for the objects.
```
```suggestion
            identifier: Some unique key for the object.
```

michamos commented on inspire_disambiguation/core/helpers.py:
```diff
@@ -0,0 +1,350 @@
+
+    @classmethod
+    def build(cls, identifier, **kwargs):
+        """Returns object from cache or builds it if it's not there.
```
```suggestion
        """Return object from cache, building it if it's not there.
```

michamos reviewed (commented):
:+1: except for minor comments. Maybe also add a test asserting that the cache is invalidated between two runs?

michamos commented:
As discussed IRL, another option is to always overwrite the cache instead of trying to be smart about it and clearing it, so you're sure you don't get stale stuff.

michamos commented on tests/integration/test_es_readers.py:
```diff
@@ -24,7 +24,8 @@
 from inspire_disambiguation.core.es.readers import get_signatures, get_input_clusters
-from inspire_disambiguation.core.data_models.publication import Publication
+from inspire_disambiguation.core.data_models.publication import Publication, \
+    PublicationCache
```
parentheses instead maybe? (use black?)

michamos reviewed (approved):
:+1: just a minor comment

monaawi committed:
disambiguation: Author disambiguation part I

*Added poetry to the project
*Removed invenio dependencies
*Upgraded to python 3
*Updated functions to use in memory dicts instead of files
*Added cli
*Added clustering per signature block

*INSPIRE-2601
